src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs

```diff
@@ -207,61 +188,79 @@ private static bool ContainsUpperCaseLetter(string pattern)
+                        // Pattern was all lowercase.  This can lead to lots of false positives.  For
+                        // example, we don't want "bin" to match "CombineUnits".  Instead, we want it
+                        // to match "BinaryOperator".  As such, make sure our match looks like it's 
+                        // starting an actual word in the candidate.  
+
+                        // Do a quick check to avoid the expensive work of having to go get the candidate
+                        // humps.  
+                        if (char.IsUpper(candidate[caseInsensitiveIndex]))
                         {
                             return new PatternMatch(PatternMatchKind.Substring, punctuationStripped,
-                                isCaseSensitive: PartStartsWith(candidate, hump, patternChunk.Text, _compareOptions),
-                                matchedSpan: GetMatchedSpan(hump.Start, patternChunk.Text.Length));
+                                isCaseSensitive: false,
+                                matchedSpan: GetMatchedSpan(caseInsensitiveIndex, patternChunk.Text.Length));
+                        }
+
+                        var candidateHumps = StringBreaker.BreakIntoWordParts(candidate);
+                        candidateHumpsOpt = candidateHumps;
+
+                        for (int i = 0, n = candidateHumps.GetCount(); i < n; i++)
+                        {
+                            var hump = TextSpan.FromBounds(candidateHumps[i].Start, candidateLength);
+                            if (PartStartsWith(candidate, hump, patternChunk.Text, CompareOptions.IgnoreCase))
+                            {
+                                return new PatternMatch(PatternMatchKind.Substring, punctuationStripped,
+                                    isCaseSensitive: PartStartsWith(candidate, hump, patternChunk.Text, CompareOptions.None),
+                                    matchedSpan: GetMatchedSpan(hump.Start, patternChunk.Text.Length));
+                            }
                         }
                     }
                 }
-            }
 
-            // Didn't have an exact/prefix match, or a high enough quality substring match.
-            // See if we can find a camel case match.  
-            return TryCamelCaseMatch(
-                candidate, patternChunk, punctuationStripped, patternIsLowercase);
+                // Didn't have an exact/prefix match, or a high enough quality substring match.
+                // See if we can find a camel case match.
+                if (candidateHumpsOpt == null)
+                {
+                    candidateHumpsOpt = StringBreaker.BreakIntoWordParts(candidate);
+                }
+
+                return TryCamelCaseMatch(
+                    candidate, patternChunk, punctuationStripped, patternIsLowercase, candidateHumpsOpt.Value);
+            }
+            finally
+            {
+                if (candidateHumpsOpt.HasValue)
```

Comment:
:bulb: Either use `HasValue` or comparisons with null, but not a mix

:bulb: You can actually use the `?.` operator on a nullable value type, reducing the finally block to this:

```csharp
candidateHumpsOpt?.Dispose();
```